### PR TITLE
Halloween Event!

### PR DESCRIPTION
**Pull request by MaxIsJoe**
https://youtu.be/hbs4Ei--Gk4

### Notes:

- Gillies, you don't need to use `AddComponent` when the component already exists in the prefab :)
- Because this is a manager that doesn't get destroyed, we do not need to remove handlers when the round ends. Do that and you can no longer assign these back.
- Leave the `Logger.Log` messages as they are useful.
- This event runs from the start of this month to the end of october.

# Changelog:

CL: [New] New Halloween timed event has been added. Enjoy your time with ghosts and free fruit baskets.
CL: [Fix] Fixed issue with events UI having missing references and broken spawn logic.

### CODE REVIEW

**Comment by MaxIsJoe** [comment]
Fixed all issues, it's good to go now.

**Comment by MaxIsJoe** [comment]
nevermind for some reason it doesn't want to work on the second round

**Comment by MaxIsJoe** [comment]
ok, now its fixed.

**Comment by MaxIsJoe** [comment]
Also, happy 9250th PR.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+				Logger.LogError("[TimedEvent/Halloween/Ghosts] - No ghosts assigned to spawn them!!", Category.Event);
+				yield break;
+			}
+
+			isRunning = true;
+
+			while (isRunning)
+			{
+				yield return WaitFor.Seconds(WAIT_TIME_BEFORE_HAUNTS);
+				if (PlayerList.Instance.ConnectionCount == 0) continue;
```
Comment (AtnerNT) [inline UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs:37]: Is this supposed to be != 0? Or can ghosts only spawn if no one is online?

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs [hunk 2]
```diff
@@ -10,26 +11,62 @@ namespace Managers
...
 	/// TODO : Allow admins to create and save events for their sever
-	public class TimedEventsManager : SingletonManager<TimedEventsManager>
+	public class TimedEventsManager : SingletonManager<TimedEventsManager>, IInitialise
 	{
 		[SerializeField] private List<TimedGameEventSO> events;
 		private List<TimedGameEventSO> activeEvents = new List<TimedGameEventSO>();
 
 		public List<TimedGameEventSO> ActiveEvents => activeEvents;
 
+		public InitialisationSystems Subsystem { get; }
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:21]: make sure you add this to the start-up list

**Review by Bod9001 — DISMISSED** [review]
llooks good

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:21]: I already did

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs [hunk 3]
```diff
@@ -10,26 +11,62 @@ namespace Managers
...
+				StartCoroutine(timedEvent.EventStart());
+			}
+		}
+
+		private void EndActiveEvents()
+		{
+			Logger.Log("[Subsystems/TimedEvents] - Stopping timed events.");
+			foreach (var timedEvent in activeEvents)
+			{
+				StartCoroutine(timedEvent.OnRoundEnd());
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:58]: Why are these coroutines?

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+using Random = UnityEngine.Random;
+
+namespace ScriptableObjects.TimedGameEvents
+{
+	[CreateAssetMenu(fileName = "HalloweenGhostsTimedEvent", menuName = "ScriptableObjects/Events/TimedGameEvents/HalloweenGhosts")]
+	public class HalloweenGhosts : TimedGameEventSO
+	{
+		[SerializeField] private List<GameObject> horrorsToSpawn;
+		[SerializeField] private SpawnPointCategory spawnPointCategory = SpawnPointCategory.MaintSpawns;
+		[SerializeField] private Vector2 randomSpawnCount = new Vector2(1, 5);
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: `Vector2` is used for positions. The UI will show an X and Y label, which definitely isn't clear.

Can we use a `randomSpawnMin` and `randomSpawnMax` instead?

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			Array catag = Enum.GetValues(typeof(SpawnPointCategory));
+			var newSpawnPoint = (SpawnPointCategory)catag.GetValue(Random.Range(0, catag.Length));
+			var previousPoint = spawnPointCategory;
+			spawnPointCategory = newSpawnPoint;
+			if(SetSpawns() == false) spawnPointCategory = previousPoint;
+		}
+
+		private void SpawnGhosts()
+		{
+			for (int i = 0; i < Random.Range(randomSpawnCount.x, randomSpawnCount.y); i++)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: Generating a random in the loop condition will update it every iteration.

Move it into a variable before the loop.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs [hunk 6]
```diff
@@ -16,11 +13,26 @@ public class TimedGameEventSO : ScriptableObject
...
+		public virtual IEnumerator OnRoundEnd()
+		{
+			yield break;
+		}
+
+		public virtual void Clean()
+		{
+			//Implement your own custom logic to clean up your scriptable objects here.
+			//This will be called when the round ends but you can call it whenever needed.
+		}
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: Could these be abstract instead? So they aren't forgotten.

**Review by Alainx277 — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: We've been using vector2 for ranged values forever now
it takes less space in the inspector, and no one has ever complained about it being "confusing" in all the time I've been here.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs (hunk 6 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: good point

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:58]: Because this is called when the subsystems load, we need access to `WaitFor` to set up things after a frame has passed or to await other coroutines and such.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs (hunk 6 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: image

You can't have an abstract method in a scriptable object

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: if you say "just make the class abstract"
image

it breaks events

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: also, some events don't need these methods

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: done

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs (hunk 6 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: Unity moment.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs (hunk 4 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: Strange but okay. How about `Vector2Int`?

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs (hunk 3 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:58]: Even when the round ends?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:58]: > Even when the round ends?

Just in case you have like hundreds of things that you want to do without freezing round end.
For example, scanning an entire matrix for objects with a specific component.

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/HalloweenGhosts.cs]: done

File: UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs (hunk 3 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/Managers/TimedEventsManager.cs:58]: Fair enough.

**Review by Alainx277 — APPROVED** [review]

File: UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs (hunk 6 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/ScriptableObjects/TimedGameEvents/TimedGameEventSO.cs]: > You can't have an abstract method in a scriptable object

not entirely true. I always do an abstract base class extending scriptable objects and then my concrete classes extending that one. Your Events break when you make it abstract because by creating assets from the abstract, you are trying to instantiate it. 

TLDR: make abstract parent and extend it with your concrete classes that will generate .asset files.
